{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/CRISTIANJULIOCESAR/DANN/blob/main/dann_cv_shaps_global.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OlJ6yuF0BjFD",
        "outputId": "7ff55117-b256-47a7-a2e9-d65fbbbbf035"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HJLu2ikTCNMG",
        "outputId": "316bc504-2bdc-444f-c70d-2d35cfe9e8ff"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch0_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch0_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch0_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch1_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch1_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch1_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch2_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch2_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch2_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch3_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch3_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch3_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch4_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch4_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch4_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch5_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch5_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch5_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch6_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch6_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch6_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/feature_extractor_fc1_epoch7_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/domain_classifier_fc2_epoch7_iter64.pkl\n",
            "Eliminado: /content/drive/MyDrive/PANCANCER FINAL/label_predictor_fc2_epoch7_iter64.pkl\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import glob\n",
        "\n",
        "def clear_previous_features(directory):\n",
        "    # Eliminar todos los archivos .pkl en el directorio especificado\n",
        "    files = glob.glob(os.path.join(directory, '*.pkl'))\n",
        "    for f in files:\n",
        "        try:\n",
        "            os.remove(f)\n",
        "            print(f'Eliminado: {f}')\n",
        "        except Exception as e:\n",
        "            print(f'No se pudo eliminar {f}. Error: {str(e)}')\n",
        "\n",
        "# Directorio donde guardas las características\n",
        "save_dir = \"/content/drive/MyDrive/PANCANCER FINAL\"\n",
        "\n",
        "# Llamada a la función para limpiar los archivos previos\n",
        "clear_previous_features(save_dir)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iDp30KkrlbdE",
        "outputId": "61af78ba-c1a0-410b-8b62-27c4cc6d59c2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Número de núcleos de CPU: 8\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "\n",
        "# Obtener el número de núcleos\n",
        "num_cores = os.cpu_count()\n",
        "print(f\"Número de núcleos de CPU: {num_cores}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m8iOpx6ykSp4",
        "outputId": "a8cdf212-036f-4798-9d59-015b0598a49b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GPU en uso: Tesla T4\n",
            "Memoria total de la GPU: 14.75 GB\n"
          ]
        }
      ],
      "source": [
        "import torch\n",
        "\n",
        "if torch.cuda.is_available():\n",
        "    gpu_name = torch.cuda.get_device_name(0)\n",
        "    print(f\"GPU en uso: {gpu_name}\")\n",
        "    print(f\"Memoria total de la GPU: {torch.cuda.get_device_properties(0).total_memory / (1024 ** 3):.2f} GB\")\n",
        "else:\n",
        "    print(\"No se encontró una GPU disponible.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "C0iwEniT3k32",
        "outputId": "59d6157e-265b-4641-ad4e-ee72e070bd7b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting umap-learn\n",
            "  Downloading umap_learn-0.5.6-py3-none-any.whl.metadata (21 kB)\n",
            "Collecting shap\n",
            "  Downloading shap-0.46.0-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (24 kB)\n",
            "Requirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.10/dist-packages (from umap-learn) (1.26.4)\n",
            "Requirement already satisfied: scipy>=1.3.1 in /usr/local/lib/python3.10/dist-packages (from umap-learn) (1.13.1)\n",
            "Requirement already satisfied: scikit-learn>=0.22 in /usr/local/lib/python3.10/dist-packages (from umap-learn) (1.3.2)\n",
            "Requirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.10/dist-packages (from umap-learn) (0.60.0)\n",
            "Collecting pynndescent>=0.5 (from umap-learn)\n",
            "  Downloading pynndescent-0.5.13-py3-none-any.whl.metadata (6.8 kB)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from umap-learn) (4.66.5)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from shap) (2.1.4)\n",
            "Requirement already satisfied: packaging>20.9 in /usr/local/lib/python3.10/dist-packages (from shap) (24.1)\n",
            "Collecting slicer==0.0.8 (from shap)\n",
            "  Downloading slicer-0.0.8-py3-none-any.whl.metadata (4.0 kB)\n",
            "Requirement already satisfied: cloudpickle in /usr/local/lib/python3.10/dist-packages (from shap) (2.2.1)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.51.2->umap-learn) (0.43.0)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.10/dist-packages (from pynndescent>=0.5->umap-learn) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.22->umap-learn) (3.5.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->shap) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->shap) (2024.1)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas->shap) (2024.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas->shap) (1.16.0)\n",
            "Downloading umap_learn-0.5.6-py3-none-any.whl (85 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.7/85.7 kB\u001b[0m \u001b[31m8.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading shap-0.46.0-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (540 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m540.1/540.1 kB\u001b[0m \u001b[31m33.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading slicer-0.0.8-py3-none-any.whl (15 kB)\n",
            "Downloading pynndescent-0.5.13-py3-none-any.whl (56 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.9/56.9 kB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: slicer, shap, pynndescent, umap-learn\n",
            "Successfully installed pynndescent-0.5.13 shap-0.46.0 slicer-0.0.8 umap-learn-0.5.6\n",
            "/bin/bash: line 1: pytorch-lamb: command not found\n"
          ]
        }
      ],
      "source": [
        "!pip install umap-learn shap\n",
        "!pytorch-lamb"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "Vj50wHuU5w-S",
        "outputId": "aece31a1-750b-48f8-ba6f-38cc7e954125"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                      Unnamed: 0  \\\n",
              "file_id                                            \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        7001   \n",
              "dd74434f-300f-4798-8342-aad31b192e12        8001   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        1001   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd        2001   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939        5001   \n",
              "...                                          ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        3999   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e        8000   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        9000   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b        4000   \n",
              "a512642f-b720-4efb-a4e9-49b746980809       11000   \n",
              "\n",
              "                                                           data_type  \\\n",
              "file_id                                                                \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  Gene Expression Quantification   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  Gene Expression Quantification   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  Gene Expression Quantification   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  Gene Expression Quantification   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  Gene Expression Quantification   \n",
              "...                                                              ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  Gene Expression Quantification   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  Gene Expression Quantification   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  Gene Expression Quantification   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  Gene Expression Quantification   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  Gene Expression Quantification   \n",
              "\n",
              "                                                      updated_datetime  \\\n",
              "file_id                                                                  \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  2016-05-26T20:51:28.753650-05:00   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  2016-05-26T21:04:47.187859-05:00   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  2016-05-26T21:04:48.015693-05:00   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  2016-05-26T21:04:48.692729-05:00   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  2016-05-26T21:04:49.266949-05:00   \n",
              "...                                                                ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  2016-06-01T23:31:30.205560-05:00   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  2016-06-01T23:31:30.867514-05:00   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  2016-06-01T23:31:31.468195-05:00   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  2016-06-01T23:31:32.108278-05:00   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  2016-06-01T23:31:32.780942-05:00   \n",
              "\n",
              "                                                                              file_name  \\\n",
              "file_id                                                                                   \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  1db2cfc5-346d-4398-9dd4-8b92435059fd.htseq.cou...   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  5f4cf592-e135-43ff-9c11-bcba462626d3.htseq.cou...   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  aa4a76e4-5348-479c-8beb-3e4e34e18ce9.htseq.cou...   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  0c172934-64ff-44dc-bdba-4c738f5692ea.htseq.cou...   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  3025ad45-1d29-4c05-b90f-3d4a348c82c6.htseq.cou...   \n",
              "...                                                                                 ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  1ef78255-1b5c-42af-aa81-128c1b009114.htseq.cou...   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  d07c7238-807f-48b1-bf06-deaa6200f4aa.htseq.cou...   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  3c5b65ad-998a-4124-814f-760de89f05d0.htseq.cou...   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  9146cb7b-e388-41b4-aa5f-a9a571abdc40.htseq.cou...   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  4690efed-5d11-4191-addb-8d290ddc063a.htseq.cou...   \n",
              "\n",
              "                                                                md5sum  \\\n",
              "file_id                                                                  \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  bff9318f04ad067bccc0fbd2d31b45f1   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  92cf7b5b7f0a0f5dfa02919c586b72c7   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  503fa39eb3a770b869ab0c4297fd0d43   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  c61b88a22488a1a0d2fb3e7739db0b32   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  d279845ef0505630cf5a2ee4b1be5cac   \n",
              "...                                                                ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  4e139effd7c17a3b6f8c8a2dcf1ba2f5   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  b07dddd077d9db4f26b3e2ae066555c4   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  2046f591b06fa9eafb79d667f7942795   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  3f9eb3df2d87fa3cb50dc7eec6766f2d   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  d1a9a8b8095b01769048904ac8379037   \n",
              "\n",
              "                                                data_category  \\\n",
              "file_id                                                         \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  Transcriptome Profiling   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  Transcriptome Profiling   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  Transcriptome Profiling   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  Transcriptome Profiling   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  Transcriptome Profiling   \n",
              "...                                                       ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  Transcriptome Profiling   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  Transcriptome Profiling   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  Transcriptome Profiling   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  Transcriptome Profiling   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  Transcriptome Profiling   \n",
              "\n",
              "                                     experimental_strategy project  \\\n",
              "file_id                                                              \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de               RNA-Seq    TCGA   \n",
              "dd74434f-300f-4798-8342-aad31b192e12               RNA-Seq    TCGA   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96               RNA-Seq    TCGA   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd               RNA-Seq    TCGA   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939               RNA-Seq    TCGA   \n",
              "...                                                    ...     ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004               RNA-Seq    TCGA   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e               RNA-Seq    TCGA   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922               RNA-Seq    TCGA   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b               RNA-Seq    TCGA   \n",
              "a512642f-b720-4efb-a4e9-49b746980809               RNA-Seq    TCGA   \n",
              "\n",
              "                                                               sample_uuid  \\\n",
              "file_id                                                                      \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  9259e9ee-7279-4b62-8512-509cb705029c   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  bccb7f9d-8dde-45b7-a50d-6c7233239ceb   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  03877fe0-b0d6-4800-be46-b45c70350e48   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  8d3e521b-6df4-4633-bb44-7647ded94130   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  8fef5013-4e57-4043-8770-a56135aeac9b   \n",
              "...                                                                    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  f03d9b83-a4d7-4e77-abaa-5c3dbee0bcf3   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  d90a2617-bf0b-453f-a483-ceac82781a6c   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  bf8e8426-dc44-40b0-8459-ad1dbaf9ba5a   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  421c391a-33c7-4a6d-aecb-9a1cb4e8a2d9   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  a418a2e8-6a8d-4e7d-8c43-f09b521e51d8   \n",
              "\n",
              "                                        sample_barcode  ...  panel  \\\n",
              "file_id                                                 ...          \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de  TCGA-DD-AAVP-01A  ...   LIHC   \n",
              "dd74434f-300f-4798-8342-aad31b192e12  TCGA-KK-A7B2-01A  ...   PRAD   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96  TCGA-DC-6158-01A  ...   READ   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd  TCGA-DD-A4NP-01A  ...   LIHC   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  TCGA-HQ-A5ND-01A  ...   BLCA   \n",
              "...                                                ...  ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004  TCGA-BP-4790-01A  ...   KIRC   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  TCGA-N9-A4Q4-01A  ...    UCS   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  TCGA-RY-A847-01A  ...    LGG   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  TCGA-AB-2881-03A  ...   LAML   \n",
              "a512642f-b720-4efb-a4e9-49b746980809  TCGA-DJ-A4V0-01A  ...   THCA   \n",
              "\n",
              "                                                                              histology  \\\n",
              "file_id                                                                                   \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de                           Hepatocellular Carcinoma   \n",
              "dd74434f-300f-4798-8342-aad31b192e12                Prostate Adenocarcinoma Acinar Type   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96                              Rectal Adenocarcinoma   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd                           Hepatocellular Carcinoma   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939  Muscle invasive urothelial carcinoma (pT2 or a...   \n",
              "...                                                                                 ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004                  Kidney Clear Cell Renal Carcinoma   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  Uterine Carcinosarcoma/ Malignant Mixed Muller...   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922                                  Oligodendroglioma   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b                                                NaN   \n",
              "a512642f-b720-4efb-a4e9-49b746980809      Thyroid Papillary Carcinoma - Classical/usual   \n",
              "\n",
              "                                                 tissue_site     stage    T  \\\n",
              "file_id                                                                       \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de                   Liver   Stage I   T1   \n",
              "dd74434f-300f-4798-8342-aad31b192e12                Prostate       NaN  T3a   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96                  Rectum   Stage I   T2   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd                   Liver   Stage I   T1   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939                 Bladder  Stage IV  T3b   \n",
              "...                                                      ...       ...  ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004                  Kidney   Stage I  T1a   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e                  Uterus       NaN  NaN   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  Central nervous system       NaN  NaN   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b             Bone Marrow       NaN  NaN   \n",
              "a512642f-b720-4efb-a4e9-49b746980809                 Thyroid   Stage I   T2   \n",
              "\n",
              "                                        N    M residual_tumor  \\\n",
              "file_id                                                         \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de   N0   M0             R0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12   N1  NaN             R0   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96   N0   M0             R0   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd   N0   M0             R0   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939   N1   M0            NaN   \n",
              "...                                   ...  ...            ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004   NX   M0            NaN   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e  NaN  NaN             R0   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922  NaN  NaN            NaN   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b  NaN  NaN            NaN   \n",
              "a512642f-b720-4efb-a4e9-49b746980809   N0   M0             R0   \n",
              "\n",
              "                                     new_tumor_events follow_ups  \n",
              "file_id                                                           \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de              2.0        1.0  \n",
              "dd74434f-300f-4798-8342-aad31b192e12              2.0        1.0  \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96              3.0        2.0  \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd              3.0        2.0  \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939              2.0        1.0  \n",
              "...                                               ...        ...  \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004              2.0        1.0  \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e              1.0        0.0  \n",
              "3d93b2ae-7730-4914-af23-ff3554752922              2.0        1.0  \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b              0.0        0.0  \n",
              "a512642f-b720-4efb-a4e9-49b746980809              3.0        2.0  \n",
              "\n",
              "[11035 rows x 27 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b39c4672-0a65-41a8-8439-6acf258a4de7\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>data_type</th>\n",
              "      <th>updated_datetime</th>\n",
              "      <th>file_name</th>\n",
              "      <th>md5sum</th>\n",
              "      <th>data_category</th>\n",
              "      <th>experimental_strategy</th>\n",
              "      <th>project</th>\n",
              "      <th>sample_uuid</th>\n",
              "      <th>sample_barcode</th>\n",
              "      <th>...</th>\n",
              "      <th>panel</th>\n",
              "      <th>histology</th>\n",
              "      <th>tissue_site</th>\n",
              "      <th>stage</th>\n",
              "      <th>T</th>\n",
              "      <th>N</th>\n",
              "      <th>M</th>\n",
              "      <th>residual_tumor</th>\n",
              "      <th>new_tumor_events</th>\n",
              "      <th>follow_ups</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>file_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>9fe7d0f4-866d-490f-b2cb-d9be7da889de</th>\n",
              "      <td>7001</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-05-26T20:51:28.753650-05:00</td>\n",
              "      <td>1db2cfc5-346d-4398-9dd4-8b92435059fd.htseq.cou...</td>\n",
              "      <td>bff9318f04ad067bccc0fbd2d31b45f1</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>9259e9ee-7279-4b62-8512-509cb705029c</td>\n",
              "      <td>TCGA-DD-AAVP-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>LIHC</td>\n",
              "      <td>Hepatocellular Carcinoma</td>\n",
              "      <td>Liver</td>\n",
              "      <td>Stage I</td>\n",
              "      <td>T1</td>\n",
              "      <td>N0</td>\n",
              "      <td>M0</td>\n",
              "      <td>R0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dd74434f-300f-4798-8342-aad31b192e12</th>\n",
              "      <td>8001</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-05-26T21:04:47.187859-05:00</td>\n",
              "      <td>5f4cf592-e135-43ff-9c11-bcba462626d3.htseq.cou...</td>\n",
              "      <td>92cf7b5b7f0a0f5dfa02919c586b72c7</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>bccb7f9d-8dde-45b7-a50d-6c7233239ceb</td>\n",
              "      <td>TCGA-KK-A7B2-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>PRAD</td>\n",
              "      <td>Prostate Adenocarcinoma Acinar Type</td>\n",
              "      <td>Prostate</td>\n",
              "      <td>NaN</td>\n",
              "      <td>T3a</td>\n",
              "      <td>N1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>07db9e9d-059d-4e6c-b483-81517a492b96</th>\n",
              "      <td>1001</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-05-26T21:04:48.015693-05:00</td>\n",
              "      <td>aa4a76e4-5348-479c-8beb-3e4e34e18ce9.htseq.cou...</td>\n",
              "      <td>503fa39eb3a770b869ab0c4297fd0d43</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>03877fe0-b0d6-4800-be46-b45c70350e48</td>\n",
              "      <td>TCGA-DC-6158-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>READ</td>\n",
              "      <td>Rectal Adenocarcinoma</td>\n",
              "      <td>Rectum</td>\n",
              "      <td>Stage I</td>\n",
              "      <td>T2</td>\n",
              "      <td>N0</td>\n",
              "      <td>M0</td>\n",
              "      <td>R0</td>\n",
              "      <td>3.0</td>\n",
              "      <td>2.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47ee860a-fd18-48f2-bfe7-4e19391950bd</th>\n",
              "      <td>2001</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-05-26T21:04:48.692729-05:00</td>\n",
              "      <td>0c172934-64ff-44dc-bdba-4c738f5692ea.htseq.cou...</td>\n",
              "      <td>c61b88a22488a1a0d2fb3e7739db0b32</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>8d3e521b-6df4-4633-bb44-7647ded94130</td>\n",
              "      <td>TCGA-DD-A4NP-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>LIHC</td>\n",
              "      <td>Hepatocellular Carcinoma</td>\n",
              "      <td>Liver</td>\n",
              "      <td>Stage I</td>\n",
              "      <td>T1</td>\n",
              "      <td>N0</td>\n",
              "      <td>M0</td>\n",
              "      <td>R0</td>\n",
              "      <td>3.0</td>\n",
              "      <td>2.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>e322b0e5-458a-4cbf-a64f-9b348e413939</th>\n",
              "      <td>5001</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-05-26T21:04:49.266949-05:00</td>\n",
              "      <td>3025ad45-1d29-4c05-b90f-3d4a348c82c6.htseq.cou...</td>\n",
              "      <td>d279845ef0505630cf5a2ee4b1be5cac</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>8fef5013-4e57-4043-8770-a56135aeac9b</td>\n",
              "      <td>TCGA-HQ-A5ND-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>BLCA</td>\n",
              "      <td>Muscle invasive urothelial carcinoma (pT2 or a...</td>\n",
              "      <td>Bladder</td>\n",
              "      <td>Stage IV</td>\n",
              "      <td>T3b</td>\n",
              "      <td>N1</td>\n",
              "      <td>M0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62cb41b1-8960-4fa2-8f74-38dd174af004</th>\n",
              "      <td>3999</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-06-01T23:31:30.205560-05:00</td>\n",
              "      <td>1ef78255-1b5c-42af-aa81-128c1b009114.htseq.cou...</td>\n",
              "      <td>4e139effd7c17a3b6f8c8a2dcf1ba2f5</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>f03d9b83-a4d7-4e77-abaa-5c3dbee0bcf3</td>\n",
              "      <td>TCGA-BP-4790-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>KIRC</td>\n",
              "      <td>Kidney Clear Cell Renal Carcinoma</td>\n",
              "      <td>Kidney</td>\n",
              "      <td>Stage I</td>\n",
              "      <td>T1a</td>\n",
              "      <td>NX</td>\n",
              "      <td>M0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>673a85f3-ba5e-448a-acee-dfec77b5238e</th>\n",
              "      <td>8000</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-06-01T23:31:30.867514-05:00</td>\n",
              "      <td>d07c7238-807f-48b1-bf06-deaa6200f4aa.htseq.cou...</td>\n",
              "      <td>b07dddd077d9db4f26b3e2ae066555c4</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>d90a2617-bf0b-453f-a483-ceac82781a6c</td>\n",
              "      <td>TCGA-N9-A4Q4-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>UCS</td>\n",
              "      <td>Uterine Carcinosarcoma/ Malignant Mixed Muller...</td>\n",
              "      <td>Uterus</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3d93b2ae-7730-4914-af23-ff3554752922</th>\n",
              "      <td>9000</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-06-01T23:31:31.468195-05:00</td>\n",
              "      <td>3c5b65ad-998a-4124-814f-760de89f05d0.htseq.cou...</td>\n",
              "      <td>2046f591b06fa9eafb79d667f7942795</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>bf8e8426-dc44-40b0-8459-ad1dbaf9ba5a</td>\n",
              "      <td>TCGA-RY-A847-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>LGG</td>\n",
              "      <td>Oligodendroglioma</td>\n",
              "      <td>Central nervous system</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>85a0b88f-2b89-4117-9e04-b81fad7f2e1b</th>\n",
              "      <td>4000</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-06-01T23:31:32.108278-05:00</td>\n",
              "      <td>9146cb7b-e388-41b4-aa5f-a9a571abdc40.htseq.cou...</td>\n",
              "      <td>3f9eb3df2d87fa3cb50dc7eec6766f2d</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>421c391a-33c7-4a6d-aecb-9a1cb4e8a2d9</td>\n",
              "      <td>TCGA-AB-2881-03A</td>\n",
              "      <td>...</td>\n",
              "      <td>LAML</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Bone Marrow</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a512642f-b720-4efb-a4e9-49b746980809</th>\n",
              "      <td>11000</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>2016-06-01T23:31:32.780942-05:00</td>\n",
              "      <td>4690efed-5d11-4191-addb-8d290ddc063a.htseq.cou...</td>\n",
              "      <td>d1a9a8b8095b01769048904ac8379037</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>RNA-Seq</td>\n",
              "      <td>TCGA</td>\n",
              "      <td>a418a2e8-6a8d-4e7d-8c43-f09b521e51d8</td>\n",
              "      <td>TCGA-DJ-A4V0-01A</td>\n",
              "      <td>...</td>\n",
              "      <td>THCA</td>\n",
              "      <td>Thyroid Papillary Carcinoma - Classical/usual</td>\n",
              "      <td>Thyroid</td>\n",
              "      <td>Stage I</td>\n",
              "      <td>T2</td>\n",
              "      <td>N0</td>\n",
              "      <td>M0</td>\n",
              "      <td>R0</td>\n",
              "      <td>3.0</td>\n",
              "      <td>2.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>11035 rows × 27 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b39c4672-0a65-41a8-8439-6acf258a4de7')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b39c4672-0a65-41a8-8439-6acf258a4de7 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b39c4672-0a65-41a8-8439-6acf258a4de7');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-80b1dd94-432d-462a-9acf-e8dfe3e950fd\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-80b1dd94-432d-462a-9acf-e8dfe3e950fd')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-80b1dd94-432d-462a-9acf-e8dfe3e950fd button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_0b614eb8-da7a-4587-bf8d-d08d37ac430f\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('clinicos')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_0b614eb8-da7a-4587-bf8d-d08d37ac430f button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('clinicos');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "clinicos"
            }
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "import pandas as pd\n",
        "clinicos = pd.read_pickle('/content/drive/MyDrive/pan_cancer_diner/final/Pan Cancer/Validados/picklesnuevos/meta_sinFPPP.pkl')\n",
        "clinicos"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 999
        },
        "id": "LuSIpxJA61CD",
        "outputId": "3be9c279-8dea-4bbe-d853-a0ebc0d50923"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0                                     5S_rRNA  5_8S_rRNA   7SK   A1BG  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        0          0    79   1055   \n",
              "dd74434f-300f-4798-8342-aad31b192e12        1          0  1250      8   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        0          0    41      2   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd        0          0   111  16482   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939        2          0    22      2   \n",
              "...                                       ...        ...   ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        3          0    52     14   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e        3          0   169     40   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        0          1  3961     10   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b        9          1  1003     90   \n",
              "a512642f-b720-4efb-a4e9-49b746980809        2          0    98     41   \n",
              "\n",
              "0                                     A1BG-AS1   A1CF     A2M  A2M-AS1  A2ML1  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        12   3291    6614       11      0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12        51      3   24510       55    120   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        35   1618   21227       29      1   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd       328  10632  159168      124      1   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939         0      2    7488       18   1714   \n",
              "...                                        ...    ...     ...      ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        76   2971  111008       92      6   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e       142      4   25243       27    277   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        54      3   32550       26    289   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b       272      0     331       96     55   \n",
              "a512642f-b720-4efb-a4e9-49b746980809       154      1   36537      107     18   \n",
              "\n",
              "0                                     A2ML1-AS1  ...  ZYG11A  ZYG11AP1  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de          0  ...       5         0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12          0  ...      73         0   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96          0  ...       5         0   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd          0  ...      66         0   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939          0  ...      52         0   \n",
              "...                                         ...  ...     ...       ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004          0  ...      28         0   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e          0  ...     172         0   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922          1  ...       2         0   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b          0  ...      10         0   \n",
              "a512642f-b720-4efb-a4e9-49b746980809          0  ...     210         0   \n",
              "\n",
              "0                                     ZYG11B    ZYX  ZYXP1  ZZEF1  ZZZ3  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de     557   3130      0    314   469   \n",
              "dd74434f-300f-4798-8342-aad31b192e12     963  10619      0   1656  1146   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96    2136  13793      0   3667  2641   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd    3417    887      0   1822  1891   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939    2155  13427      0   1404  2215   \n",
              "...                                      ...    ...    ...    ...   ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004    3969  11102      0   4381  2584   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e    2717  14657      0   1090  2165   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922    2568   4216      0   3239   908   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b    1611  19604      0   7247  3970   \n",
              "a512642f-b720-4efb-a4e9-49b746980809    4057   8229      0   5007  3273   \n",
              "\n",
              "0                                     hsa-mir-1253  hsa-mir-423  snoZ196  \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de             0            0        3  \n",
              "dd74434f-300f-4798-8342-aad31b192e12             0            0        3  \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96             0            0        3  \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd             0            0        1  \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939             0            0        4  \n",
              "...                                            ...          ...      ...  \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004             0            0        1  \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e             0            0        3  \n",
              "3d93b2ae-7730-4914-af23-ff3554752922             0            0        1  \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b             0            0       13  \n",
              "a512642f-b720-4efb-a4e9-49b746980809             0            0        7  \n",
              "\n",
              "[10310 rows x 39979 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-624bd742-9827-416e-9a19-47f72abcd217\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>5S_rRNA</th>\n",
              "      <th>5_8S_rRNA</th>\n",
              "      <th>7SK</th>\n",
              "      <th>A1BG</th>\n",
              "      <th>A1BG-AS1</th>\n",
              "      <th>A1CF</th>\n",
              "      <th>A2M</th>\n",
              "      <th>A2M-AS1</th>\n",
              "      <th>A2ML1</th>\n",
              "      <th>A2ML1-AS1</th>\n",
              "      <th>...</th>\n",
              "      <th>ZYG11A</th>\n",
              "      <th>ZYG11AP1</th>\n",
              "      <th>ZYG11B</th>\n",
              "      <th>ZYX</th>\n",
              "      <th>ZYXP1</th>\n",
              "      <th>ZZEF1</th>\n",
              "      <th>ZZZ3</th>\n",
              "      <th>hsa-mir-1253</th>\n",
              "      <th>hsa-mir-423</th>\n",
              "      <th>snoZ196</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>9fe7d0f4-866d-490f-b2cb-d9be7da889de</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>79</td>\n",
              "      <td>1055</td>\n",
              "      <td>12</td>\n",
              "      <td>3291</td>\n",
              "      <td>6614</td>\n",
              "      <td>11</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>557</td>\n",
              "      <td>3130</td>\n",
              "      <td>0</td>\n",
              "      <td>314</td>\n",
              "      <td>469</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dd74434f-300f-4798-8342-aad31b192e12</th>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1250</td>\n",
              "      <td>8</td>\n",
              "      <td>51</td>\n",
              "      <td>3</td>\n",
              "      <td>24510</td>\n",
              "      <td>55</td>\n",
              "      <td>120</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>73</td>\n",
              "      <td>0</td>\n",
              "      <td>963</td>\n",
              "      <td>10619</td>\n",
              "      <td>0</td>\n",
              "      <td>1656</td>\n",
              "      <td>1146</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>07db9e9d-059d-4e6c-b483-81517a492b96</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>41</td>\n",
              "      <td>2</td>\n",
              "      <td>35</td>\n",
              "      <td>1618</td>\n",
              "      <td>21227</td>\n",
              "      <td>29</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>2136</td>\n",
              "      <td>13793</td>\n",
              "      <td>0</td>\n",
              "      <td>3667</td>\n",
              "      <td>2641</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47ee860a-fd18-48f2-bfe7-4e19391950bd</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>111</td>\n",
              "      <td>16482</td>\n",
              "      <td>328</td>\n",
              "      <td>10632</td>\n",
              "      <td>159168</td>\n",
              "      <td>124</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>66</td>\n",
              "      <td>0</td>\n",
              "      <td>3417</td>\n",
              "      <td>887</td>\n",
              "      <td>0</td>\n",
              "      <td>1822</td>\n",
              "      <td>1891</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>e322b0e5-458a-4cbf-a64f-9b348e413939</th>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>22</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>7488</td>\n",
              "      <td>18</td>\n",
              "      <td>1714</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>52</td>\n",
              "      <td>0</td>\n",
              "      <td>2155</td>\n",
              "      <td>13427</td>\n",
              "      <td>0</td>\n",
              "      <td>1404</td>\n",
              "      <td>2215</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62cb41b1-8960-4fa2-8f74-38dd174af004</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>52</td>\n",
              "      <td>14</td>\n",
              "      <td>76</td>\n",
              "      <td>2971</td>\n",
              "      <td>111008</td>\n",
              "      <td>92</td>\n",
              "      <td>6</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>28</td>\n",
              "      <td>0</td>\n",
              "      <td>3969</td>\n",
              "      <td>11102</td>\n",
              "      <td>0</td>\n",
              "      <td>4381</td>\n",
              "      <td>2584</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>673a85f3-ba5e-448a-acee-dfec77b5238e</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>169</td>\n",
              "      <td>40</td>\n",
              "      <td>142</td>\n",
              "      <td>4</td>\n",
              "      <td>25243</td>\n",
              "      <td>27</td>\n",
              "      <td>277</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>172</td>\n",
              "      <td>0</td>\n",
              "      <td>2717</td>\n",
              "      <td>14657</td>\n",
              "      <td>0</td>\n",
              "      <td>1090</td>\n",
              "      <td>2165</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3d93b2ae-7730-4914-af23-ff3554752922</th>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>3961</td>\n",
              "      <td>10</td>\n",
              "      <td>54</td>\n",
              "      <td>3</td>\n",
              "      <td>32550</td>\n",
              "      <td>26</td>\n",
              "      <td>289</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2568</td>\n",
              "      <td>4216</td>\n",
              "      <td>0</td>\n",
              "      <td>3239</td>\n",
              "      <td>908</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>85a0b88f-2b89-4117-9e04-b81fad7f2e1b</th>\n",
              "      <td>9</td>\n",
              "      <td>1</td>\n",
              "      <td>1003</td>\n",
              "      <td>90</td>\n",
              "      <td>272</td>\n",
              "      <td>0</td>\n",
              "      <td>331</td>\n",
              "      <td>96</td>\n",
              "      <td>55</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>1611</td>\n",
              "      <td>19604</td>\n",
              "      <td>0</td>\n",
              "      <td>7247</td>\n",
              "      <td>3970</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>13</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a512642f-b720-4efb-a4e9-49b746980809</th>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>98</td>\n",
              "      <td>41</td>\n",
              "      <td>154</td>\n",
              "      <td>1</td>\n",
              "      <td>36537</td>\n",
              "      <td>107</td>\n",
              "      <td>18</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>210</td>\n",
              "      <td>0</td>\n",
              "      <td>4057</td>\n",
              "      <td>8229</td>\n",
              "      <td>0</td>\n",
              "      <td>5007</td>\n",
              "      <td>3273</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10310 rows × 39979 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-624bd742-9827-416e-9a19-47f72abcd217')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-624bd742-9827-416e-9a19-47f72abcd217 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-624bd742-9827-416e-9a19-47f72abcd217');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-ad52ab23-18e0-4dc9-9b84-88e982120da6\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ad52ab23-18e0-4dc9-9b84-88e982120da6')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-ad52ab23-18e0-4dc9-9b84-88e982120da6 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_f3df2381-d149-42e0-ac9c-fcf12ac8edbf\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_f3df2381-d149-42e0-ac9c-fcf12ac8edbf button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "df = pd.read_pickle('/content/drive/MyDrive/pan_cancer_diner/final/Pan Cancer/Validados/picklesnuevos/df_sinFPPP_2.pkl')\n",
        "df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "x5oyeu0mRe8F"
      },
      "outputs": [],
      "source": [
        "clinicos = clinicos.dropna(subset=['vital'])\n",
        "clinicos = clinicos.dropna(subset=['panel'])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zhrpWaDJIkOa"
      },
      "outputs": [],
      "source": [
        "df['vital'] = clinicos['vital']\n",
        "df['panel'] = clinicos['panel']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tPMx64NFRjsK"
      },
      "outputs": [],
      "source": [
        "df = df.dropna(subset=['vital'])\n",
        "df = df.dropna(subset=['panel'])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "W3ji0diK96NW",
        "outputId": "a49122f1-242d-4868-8121-efad83d02593"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['LIHC', 'PRAD', 'READ', 'BLCA', 'LGG', 'BRCA', 'UCEC', 'SARC',\n",
              "       'KIRC', 'PCPG', 'DLBC', 'MESO', 'THYM', 'LUSC', 'LUAD', 'STAD',\n",
              "       'GBM', 'LAML', 'THCA', 'CESC', 'COAD', 'HNSC', 'UVM', 'SKCM',\n",
              "       'PAAD', 'OV', 'TGCT', 'KICH', 'ESCA', 'KIRP', 'ACC', 'CHOL', 'UCS'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ],
      "source": [
        "clinicos['panel'].unique()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 999
        },
        "id": "_uh4cS8CHE5K",
        "outputId": "3a59f744-2074-4e89-fa7e-28eb89028864"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0                                     5S_rRNA  5_8S_rRNA   7SK   A1BG  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        0          0    79   1055   \n",
              "dd74434f-300f-4798-8342-aad31b192e12        1          0  1250      8   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        0          0    41      2   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd        0          0   111  16482   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939        2          0    22      2   \n",
              "...                                       ...        ...   ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        3          0    52     14   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e        3          0   169     40   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        0          1  3961     10   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b        9          1  1003     90   \n",
              "a512642f-b720-4efb-a4e9-49b746980809        2          0    98     41   \n",
              "\n",
              "0                                     A1BG-AS1   A1CF     A2M  A2M-AS1  A2ML1  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        12   3291    6614       11      0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12        51      3   24510       55    120   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        35   1618   21227       29      1   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd       328  10632  159168      124      1   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939         0      2    7488       18   1714   \n",
              "...                                        ...    ...     ...      ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        76   2971  111008       92      6   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e       142      4   25243       27    277   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        54      3   32550       26    289   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b       272      0     331       96     55   \n",
              "a512642f-b720-4efb-a4e9-49b746980809       154      1   36537      107     18   \n",
              "\n",
              "0                                     A2ML1-AS1  ...  ZYG11B    ZYX  ZYXP1  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de          0  ...     557   3130      0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12          0  ...     963  10619      0   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96          0  ...    2136  13793      0   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd          0  ...    3417    887      0   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939          0  ...    2155  13427      0   \n",
              "...                                         ...  ...     ...    ...    ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004          0  ...    3969  11102      0   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e          0  ...    2717  14657      0   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922          1  ...    2568   4216      0   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b          0  ...    1611  19604      0   \n",
              "a512642f-b720-4efb-a4e9-49b746980809          0  ...    4057   8229      0   \n",
              "\n",
              "0                                     ZZEF1  ZZZ3  hsa-mir-1253  hsa-mir-423  \\\n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de    314   469             0            0   \n",
              "dd74434f-300f-4798-8342-aad31b192e12   1656  1146             0            0   \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96   3667  2641             0            0   \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd   1822  1891             0            0   \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939   1404  2215             0            0   \n",
              "...                                     ...   ...           ...          ...   \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004   4381  2584             0            0   \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e   1090  2165             0            0   \n",
              "3d93b2ae-7730-4914-af23-ff3554752922   3239   908             0            0   \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b   7247  3970             0            0   \n",
              "a512642f-b720-4efb-a4e9-49b746980809   5007  3273             0            0   \n",
              "\n",
              "0                                     snoZ196  vital  panel  \n",
              "9fe7d0f4-866d-490f-b2cb-d9be7da889de        3  Alive   LIHC  \n",
              "dd74434f-300f-4798-8342-aad31b192e12        3  Alive   PRAD  \n",
              "07db9e9d-059d-4e6c-b483-81517a492b96        3   Dead   READ  \n",
              "47ee860a-fd18-48f2-bfe7-4e19391950bd        1  Alive   LIHC  \n",
              "e322b0e5-458a-4cbf-a64f-9b348e413939        4   Dead   BLCA  \n",
              "...                                       ...    ...    ...  \n",
              "62cb41b1-8960-4fa2-8f74-38dd174af004        1   Dead   KIRC  \n",
              "673a85f3-ba5e-448a-acee-dfec77b5238e        3  Alive    UCS  \n",
              "3d93b2ae-7730-4914-af23-ff3554752922        1  Alive    LGG  \n",
              "85a0b88f-2b89-4117-9e04-b81fad7f2e1b       13  Alive   LAML  \n",
              "a512642f-b720-4efb-a4e9-49b746980809        7  Alive   THCA  \n",
              "\n",
              "[10286 rows x 39981 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0354f0bd-8c6b-40b9-936c-ac5ad26a2500\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>5S_rRNA</th>\n",
              "      <th>5_8S_rRNA</th>\n",
              "      <th>7SK</th>\n",
              "      <th>A1BG</th>\n",
              "      <th>A1BG-AS1</th>\n",
              "      <th>A1CF</th>\n",
              "      <th>A2M</th>\n",
              "      <th>A2M-AS1</th>\n",
              "      <th>A2ML1</th>\n",
              "      <th>A2ML1-AS1</th>\n",
              "      <th>...</th>\n",
              "      <th>ZYG11B</th>\n",
              "      <th>ZYX</th>\n",
              "      <th>ZYXP1</th>\n",
              "      <th>ZZEF1</th>\n",
              "      <th>ZZZ3</th>\n",
              "      <th>hsa-mir-1253</th>\n",
              "      <th>hsa-mir-423</th>\n",
              "      <th>snoZ196</th>\n",
              "      <th>vital</th>\n",
              "      <th>panel</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>9fe7d0f4-866d-490f-b2cb-d9be7da889de</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>79</td>\n",
              "      <td>1055</td>\n",
              "      <td>12</td>\n",
              "      <td>3291</td>\n",
              "      <td>6614</td>\n",
              "      <td>11</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>557</td>\n",
              "      <td>3130</td>\n",
              "      <td>0</td>\n",
              "      <td>314</td>\n",
              "      <td>469</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>Alive</td>\n",
              "      <td>LIHC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>dd74434f-300f-4798-8342-aad31b192e12</th>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1250</td>\n",
              "      <td>8</td>\n",
              "      <td>51</td>\n",
              "      <td>3</td>\n",
              "      <td>24510</td>\n",
              "      <td>55</td>\n",
              "      <td>120</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>963</td>\n",
              "      <td>10619</td>\n",
              "      <td>0</td>\n",
              "      <td>1656</td>\n",
              "      <td>1146</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>Alive</td>\n",
              "      <td>PRAD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>07db9e9d-059d-4e6c-b483-81517a492b96</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>41</td>\n",
              "      <td>2</td>\n",
              "      <td>35</td>\n",
              "      <td>1618</td>\n",
              "      <td>21227</td>\n",
              "      <td>29</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>2136</td>\n",
              "      <td>13793</td>\n",
              "      <td>0</td>\n",
              "      <td>3667</td>\n",
              "      <td>2641</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>Dead</td>\n",
              "      <td>READ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47ee860a-fd18-48f2-bfe7-4e19391950bd</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>111</td>\n",
              "      <td>16482</td>\n",
              "      <td>328</td>\n",
              "      <td>10632</td>\n",
              "      <td>159168</td>\n",
              "      <td>124</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>3417</td>\n",
              "      <td>887</td>\n",
              "      <td>0</td>\n",
              "      <td>1822</td>\n",
              "      <td>1891</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>Alive</td>\n",
              "      <td>LIHC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>e322b0e5-458a-4cbf-a64f-9b348e413939</th>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>22</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>7488</td>\n",
              "      <td>18</td>\n",
              "      <td>1714</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>2155</td>\n",
              "      <td>13427</td>\n",
              "      <td>0</td>\n",
              "      <td>1404</td>\n",
              "      <td>2215</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>Dead</td>\n",
              "      <td>BLCA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62cb41b1-8960-4fa2-8f74-38dd174af004</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>52</td>\n",
              "      <td>14</td>\n",
              "      <td>76</td>\n",
              "      <td>2971</td>\n",
              "      <td>111008</td>\n",
              "      <td>92</td>\n",
              "      <td>6</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>3969</td>\n",
              "      <td>11102</td>\n",
              "      <td>0</td>\n",
              "      <td>4381</td>\n",
              "      <td>2584</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>Dead</td>\n",
              "      <td>KIRC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>673a85f3-ba5e-448a-acee-dfec77b5238e</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>169</td>\n",
              "      <td>40</td>\n",
              "      <td>142</td>\n",
              "      <td>4</td>\n",
              "      <td>25243</td>\n",
              "      <td>27</td>\n",
              "      <td>277</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>2717</td>\n",
              "      <td>14657</td>\n",
              "      <td>0</td>\n",
              "      <td>1090</td>\n",
              "      <td>2165</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>Alive</td>\n",
              "      <td>UCS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3d93b2ae-7730-4914-af23-ff3554752922</th>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>3961</td>\n",
              "      <td>10</td>\n",
              "      <td>54</td>\n",
              "      <td>3</td>\n",
              "      <td>32550</td>\n",
              "      <td>26</td>\n",
              "      <td>289</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>2568</td>\n",
              "      <td>4216</td>\n",
              "      <td>0</td>\n",
              "      <td>3239</td>\n",
              "      <td>908</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>Alive</td>\n",
              "      <td>LGG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>85a0b88f-2b89-4117-9e04-b81fad7f2e1b</th>\n",
              "      <td>9</td>\n",
              "      <td>1</td>\n",
              "      <td>1003</td>\n",
              "      <td>90</td>\n",
              "      <td>272</td>\n",
              "      <td>0</td>\n",
              "      <td>331</td>\n",
              "      <td>96</td>\n",
              "      <td>55</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>1611</td>\n",
              "      <td>19604</td>\n",
              "      <td>0</td>\n",
              "      <td>7247</td>\n",
              "      <td>3970</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>13</td>\n",
              "      <td>Alive</td>\n",
              "      <td>LAML</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a512642f-b720-4efb-a4e9-49b746980809</th>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>98</td>\n",
              "      <td>41</td>\n",
              "      <td>154</td>\n",
              "      <td>1</td>\n",
              "      <td>36537</td>\n",
              "      <td>107</td>\n",
              "      <td>18</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>4057</td>\n",
              "      <td>8229</td>\n",
              "      <td>0</td>\n",
              "      <td>5007</td>\n",
              "      <td>3273</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "      <td>Alive</td>\n",
              "      <td>THCA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10286 rows × 39981 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0354f0bd-8c6b-40b9-936c-ac5ad26a2500')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0354f0bd-8c6b-40b9-936c-ac5ad26a2500 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0354f0bd-8c6b-40b9-936c-ac5ad26a2500');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-f75ef0b1-e18c-45ba-a5ef-2ffbe6a46824\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f75ef0b1-e18c-45ba-a5ef-2ffbe6a46824')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-f75ef0b1-e18c-45ba-a5ef-2ffbe6a46824 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_a35cb4b2-ffa6-4201-a259-db1ba6f416d7\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_a35cb4b2-ffa6-4201-a259-db1ba6f416d7 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 12
        }
      ],
      "source": [
        "df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Wj4jf1143PJW"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dIaB7MIEm3zL",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "13fb0317-d065-4219-f6ed-5a250896cdd7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Usando dispositivo: cuda\n",
            "----------------------------------------------------------------\n",
            "        Layer (type)               Output Shape         Param #\n",
            "================================================================\n",
            "            Linear-1                 [-1, 5000]     199,900,000\n",
            "       BatchNorm1d-2                 [-1, 5000]          10,000\n",
            "           Dropout-3                 [-1, 5000]               0\n",
            "  FeatureExtractor-4                 [-1, 5000]               0\n",
            "            Linear-5                 [-1, 5000]      25,005,000\n",
            "       BatchNorm1d-6                 [-1, 5000]          10,000\n",
            "           Dropout-7                 [-1, 5000]               0\n",
            "            Linear-8                 [-1, 5000]      25,005,000\n",
            "       BatchNorm1d-9                 [-1, 5000]          10,000\n",
            "          Dropout-10                 [-1, 5000]               0\n",
            "           Linear-11                    [-1, 1]           5,001\n",
            "   LabelPredictor-12                    [-1, 1]               0\n",
            "           Linear-13                 [-1, 5000]      25,005,000\n",
            "      BatchNorm1d-14                 [-1, 5000]          10,000\n",
            "          Dropout-15                 [-1, 5000]               0\n",
            "           Linear-16                 [-1, 5000]      25,005,000\n",
            "      BatchNorm1d-17                 [-1, 5000]          10,000\n",
            "          Dropout-18                 [-1, 5000]               0\n",
            "           Linear-19                   [-1, 33]         165,033\n",
            " DomainClassifier-20                   [-1, 33]               0\n",
            "================================================================\n",
            "Total params: 300,140,034\n",
            "Trainable params: 300,140,034\n",
            "Non-trainable params: 0\n",
            "----------------------------------------------------------------\n",
            "Input size (MB): 0.15\n",
            "Forward/backward pass size (MB): 0.61\n",
            "Params size (MB): 1144.94\n",
            "Estimated Total Size (MB): 1145.71\n",
            "----------------------------------------------------------------\n",
            "Epoch [1/10], Label Accuracy: 0.6105, Domain Accuracy: 0.0273\n",
            "Epoch [2/10], Label Accuracy: 0.6987, Domain Accuracy: 0.0305\n",
            "Epoch [3/10], Label Accuracy: 0.7258, Domain Accuracy: 0.0282\n",
            "Epoch [4/10], Label Accuracy: 0.7477, Domain Accuracy: 0.0312\n",
            "Epoch [5/10], Label Accuracy: 0.7670, Domain Accuracy: 0.0280\n",
            "Epoch [6/10], Label Accuracy: 0.7942, Domain Accuracy: 0.0303\n",
            "Epoch [7/10], Label Accuracy: 0.8193, Domain Accuracy: 0.0284\n",
            "Epoch [8/10], Label Accuracy: 0.8391, Domain Accuracy: 0.0272\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "OutOfMemoryError",
          "evalue": "CUDA out of memory. Tried to allocate 764.00 MiB. GPU 0 has a total capacity of 14.75 GiB of which 383.06 MiB is free. Process 11224 has 14.37 GiB memory in use. Of the allocated memory 14.17 GiB is allocated by PyTorch, and 68.33 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mOutOfMemoryError\u001b[0m                          Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-13-3c8ce8d48101>\u001b[0m in \u001b[0;36m<cell line: 285>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;31m# Actualización de los parámetros del modelo\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 314\u001b[0;31m         \u001b[0moptimizer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    315\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    316\u001b[0m         \u001b[0mpredicted_label\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlabel_preds\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0.5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/optimizer.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    482\u001b[0m                             )\n\u001b[1;32m    483\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 484\u001b[0;31m                 \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    485\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_optimizer_step_code\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    486\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/optimizer.py\u001b[0m in \u001b[0;36m_use_grad\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     87\u001b[0m             \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_grad_enabled\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefaults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"differentiable\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     88\u001b[0m             \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dynamo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph_break\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 89\u001b[0;31m             \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     90\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m             \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dynamo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph_break\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/adamw.py\u001b[0m in \u001b[0;36mstep\u001b[0;34m(self, closure)\u001b[0m\n\u001b[1;32m    225\u001b[0m             )\n\u001b[1;32m    226\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 227\u001b[0;31m             adamw(\n\u001b[0m\u001b[1;32m    228\u001b[0m                 \u001b[0mparams_with_grad\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    229\u001b[0m                 \u001b[0mgrads\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/optimizer.py\u001b[0m in \u001b[0;36mmaybe_fallback\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    159\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mdisabled_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    160\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 161\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    162\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    163\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mmaybe_fallback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/adamw.py\u001b[0m in \u001b[0;36madamw\u001b[0;34m(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps, foreach, capturable, differentiable, fused, grad_scale, found_inf, has_complex, amsgrad, beta1, beta2, lr, weight_decay, eps, maximize)\u001b[0m\n\u001b[1;32m    765\u001b[0m         \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_single_tensor_adamw\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    766\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 767\u001b[0;31m     func(\n\u001b[0m\u001b[1;32m    768\u001b[0m         \u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    769\u001b[0m         \u001b[0mgrads\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/optim/adamw.py\u001b[0m in \u001b[0;36m_multi_tensor_adamw\u001b[0;34m(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps, grad_scale, found_inf, amsgrad, beta1, beta2, lr, weight_decay, eps, maximize, capturable, differentiable, has_complex)\u001b[0m\n\u001b[1;32m    598\u001b[0m                 \u001b[0mexp_avg_sq_sqrt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_foreach_sqrt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_max_exp_avg_sqs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    599\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 600\u001b[0;31m                 \u001b[0mexp_avg_sq_sqrt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_foreach_sqrt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_exp_avg_sqs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    601\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    602\u001b[0m             \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_foreach_div_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexp_avg_sq_sqrt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbias_correction2_sqrt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mOutOfMemoryError\u001b[0m: CUDA out of memory. Tried to allocate 764.00 MiB. GPU 0 has a total capacity of 14.75 GiB of which 383.06 MiB is free. Process 11224 has 14.37 GiB memory in use. Of the allocated memory 14.17 GiB is allocated by PyTorch, and 68.33 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "import torch.nn.functional as F\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from torch.utils.data import DataLoader, TensorDataset\n",
        "import numpy as np\n",
        "import umap\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "import pickle\n",
        "from torchsummary import summary  # Importar torchsummary para imprimir el resumen del modelo\n",
        "\n",
        "# Configurar el dispositivo para GPU si está disponible, de lo contrario, usar CPU\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "print(f\"Usando dispositivo: {device}\")\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "# Dividir los datos en conjuntos de entrenamiento (80%) y prueba (20%) de forma estratificada\n",
        "X_train, X_test, Y_train, Y_test, panels_train, panels_test = train_test_split(\n",
        "    X, Y, panels, test_size=0.2, stratify=Y, random_state=42\n",
        ")\n",
        "\n",
        "# Sobrescribir X y Y con los datos de entrenamiento\n",
        "X = X_train\n",
        "Y = Y_train\n",
        "panels = panels_train\n",
        "\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Codificar las etiquetas de dominio (paneles)\n",
        "label_encoder_panel = LabelEncoder()\n",
        "panels = label_encoder_panel.fit_transform(panels)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "panels_tensor = torch.tensor(panels, dtype=torch.long).to(device)\n",
        "\n",
        "# Crear datasets y dataloaders\n",
        "batch_size = 128\n",
        "dataset = TensorDataset(X_tensor, Y_tensor, panels_tensor)\n",
        "loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)\n",
        "\n",
        "# Definición del modelo DANN mejorado\n",
        "class FeatureExtractor(nn.Module):\n",
        "    def __init__(self, input_dim):\n",
        "        super(FeatureExtractor, self).__init__()\n",
        "        self.fc1 = nn.Linear(input_dim, 5000)  # Capa única con 10,000 neuronas\n",
        "        self.bn1 = nn.BatchNorm1d(5000)\n",
        "        self.dropout1 = nn.Dropout(0.3)  # Dropout moderado\n",
        "\n",
        "        # Aplicar inicialización He a la capa\n",
        "        nn.init.kaiming_normal_(self.fc1.weight, nonlinearity='leaky_relu')\n",
        "\n",
        "    def forward(self, x, lambda_=1.0):\n",
        "        x = F.leaky_relu(self.bn1(self.fc1(x)))\n",
        "        x = self.dropout1(x)\n",
        "        return x\n",
        "\n",
        "class LabelPredictor(nn.Module):\n",
        "    def __init__(self, input_dim):\n",
        "        super(LabelPredictor, self).__init__()\n",
        "        self.fc1 = nn.Linear(input_dim, 5000)\n",
        "        self.bn1 = nn.BatchNorm1d(5000)\n",
        "        self.dropout1 = nn.Dropout(0.3)  # Dropout moderado\n",
        "        self.fc2 = nn.Linear(5000, 5000)\n",
        "        self.bn2 = nn.BatchNorm1d(5000)\n",
        "        self.dropout2 = nn.Dropout(0.3)  # Dropout moderado\n",
        "        self.fc3 = nn.Linear(5000, 1)\n",
        "\n",
        "        # Aplicar inicialización He a las capas\n",
        "        nn.init.kaiming_normal_(self.fc1.weight, nonlinearity='leaky_relu')\n",
        "        nn.init.kaiming_normal_(self.fc2.weight, nonlinearity='leaky_relu')\n",
        "        nn.init.kaiming_normal_(self.fc3.weight, nonlinearity='leaky_relu')\n",
        "\n",
        "    def forward(self, x):\n",
        "        x = F.leaky_relu(self.bn1(self.fc1(x)))\n",
        "        x = self.dropout1(x)\n",
        "        x = F.leaky_relu(self.bn2(self.fc2(x)))\n",
        "        x = self.dropout2(x)\n",
        "        x = torch.sigmoid(self.fc3(x))\n",
        "        return x\n",
        "\n",
        "class DomainClassifier(nn.Module):\n",
        "    def __init__(self, input_dim, num_domains):\n",
        "        super(DomainClassifier, self).__init__()\n",
        "        self.fc1 = nn.Linear(input_dim, 5000)\n",
        "        self.bn1 = nn.BatchNorm1d(5000)\n",
        "        self.dropout1 = nn.Dropout(0.9)  # Dropout moderado\n",
        "        self.fc2 = nn.Linear(5000, 5000)\n",
        "        self.bn2 = nn.BatchNorm1d(5000)\n",
        "        self.dropout2 = nn.Dropout(0.9)  # Dropout moderado\n",
        "        self.fc3 = nn.Linear(5000, num_domains)\n",
        "\n",
        "        # Aplicar inicialización He a las capas\n",
        "        nn.init.kaiming_normal_(self.fc1.weight, nonlinearity='leaky_relu')\n",
        "        nn.init.kaiming_normal_(self.fc2.weight, nonlinearity='leaky_relu')\n",
        "        nn.init.kaiming_normal_(self.fc3.weight, nonlinearity='leaky_relu')\n",
        "\n",
        "    def forward(self, x):\n",
        "        x = F.leaky_relu(self.bn1(self.fc1(x)))\n",
        "        x = self.dropout1(x)\n",
        "        x = F.leaky_relu(self.bn2(self.fc2(x)))\n",
        "        x = self.dropout2(x)\n",
        "        x = self.fc3(x)\n",
        "        return x\n",
        "\n",
        "class DANN(nn.Module):\n",
        "    def __init__(self, feature_dim, num_domains):\n",
        "        super(DANN, self).__init__()\n",
        "        self.feature_extractor = FeatureExtractor(feature_dim)\n",
        "        self.label_predictor = LabelPredictor(5000)\n",
        "        self.domain_classifier = DomainClassifier(5000, num_domains)\n",
        "\n",
        "    def forward(self, x, lambda_=1.0):\n",
        "        # Extracción de características\n",
        "        features = self.feature_extractor(x)\n",
        "\n",
        "        # Inversión de gradiente con lambda\n",
        "        reverse_features = ReverseLayerF.apply(features, lambda_)\n",
        "\n",
        "        # Predicción de etiquetas\n",
        "        label_output = self.label_predictor(features)\n",
        "\n",
        "        # Clasificación de dominio\n",
        "        domain_output = self.domain_classifier(reverse_features)\n",
        "\n",
        "        return label_output, domain_output, features\n",
        "\n",
        "class ReverseLayerF(torch.autograd.Function):\n",
        "    @staticmethod\n",
        "    def forward(ctx, x, lambda_):\n",
        "        ctx.lambda_ = lambda_  # Guardamos lambda para usarla en la retropropagación\n",
        "        return x.view_as(x)  # Devolvemos las características sin cambios\n",
        "\n",
        "    @staticmethod\n",
        "    def backward(ctx, grad_output):\n",
        "        # Invertimos los gradientes y los escalamos por lambda\n",
        "        output = grad_output.neg() * ctx.lambda_\n",
        "        return output, None  # Retornamos el gradiente invertido\n",
        "\n",
        "# Función para listar las capas de un módulo\n",
        "def list_module_layers(module, module_name):\n",
        "    print(f\"Capas disponibles en {module_name}:\")\n",
        "    for name, mod in module.named_children():\n",
        "        print(f'Layer name: {name}, Module: {mod}')\n",
        "\n",
        "# Función para extraer características de una capa específica de un módulo\n",
        "def extract_layer_features(model, loader, module_name, layer_name, use_features=False):\n",
        "    model.eval()\n",
        "    features = []\n",
        "    labels = []\n",
        "    domains = []\n",
        "\n",
        "    with torch.no_grad():\n",
        "        for inputs, label, domain in loader:\n",
        "            inputs, label, domain = inputs.to(device), label.to(device), domain.to(device)\n",
        "            if use_features:\n",
        "                x = model.feature_extractor(inputs)\n",
        "            else:\n",
        "                x = inputs\n",
        "            module = getattr(model, module_name)\n",
        "            for name, mod in module.named_children():\n",
        "                x = mod(x)\n",
        "                if name == layer_name:\n",
        "                    break\n",
        "            features.append(x.cpu().numpy())\n",
        "            labels.append(label.cpu().numpy())\n",
        "            domains.append(domain.cpu().numpy())\n",
        "\n",
        "    features = np.concatenate(features, axis=0)\n",
        "    labels = np.concatenate(labels, axis=0).flatten()\n",
        "    domains = np.concatenate(domains, axis=0).flatten()\n",
        "    return features, labels, domains\n",
        "\n",
        "# Función para guardar características de una capa específica de un módulo en disco\n",
        "def save_layer_features_to_disk(epoch_iter, model, loader, module_name, layer_name, use_features=False):\n",
        "    features, labels, domains = extract_layer_features(model, loader, module_name, layer_name, use_features)\n",
        "\n",
        "    # Crear un directorio para guardar las características si no existe\n",
        "    save_dir = \"/content/drive/MyDrive/PANCANCER FINAL\"\n",
        "    if not os.path.exists(save_dir):\n",
        "        os.makedirs(save_dir)\n",
        "\n",
        "    # Guardar las características en disco\n",
        "    file_name = f\"{save_dir}/{module_name}_{layer_name}_epoch{epoch_iter[0]}_iter{epoch_iter[1]}.pkl\"\n",
        "    with open(file_name, 'wb') as f:\n",
        "        pickle.dump((features, labels, domains), f)\n",
        "\n",
        "# Función para cargar características desde disco\n",
        "def load_layer_features_from_disk(epoch, iteration, module_name, layer_name):\n",
        "    file_name = f\"/content/drive/MyDrive/PANCANCER FINAL/{module_name}_{layer_name}_epoch{epoch}_iter{iteration}.pkl\"\n",
        "    if os.path.exists(file_name):\n",
        "        with open(file_name, 'rb') as f:\n",
        "            features, labels, domains = pickle.load(f)\n",
        "        return features, labels, domains\n",
        "    else:\n",
        "        print(f\"No features found for epoch {epoch}, iteration {iteration}, module {module_name}, layer {layer_name}\")\n",
        "        return None, None, None\n",
        "\n",
        "# Función para visualizar UMAP\n",
        "def plot_umap(embedding, labels, domains, title_suffix=''):\n",
        "    plt.figure(figsize=(20, 10))\n",
        "\n",
        "    plt.subplot(1, 2, 1)\n",
        "    sns.scatterplot(x=embedding[:, 0], y=embedding[:, 1], hue=domains, palette='Spectral', s=10)\n",
        "    plt.title(f'UMAP projection of the shared feature space (colored by domain) {title_suffix}')\n",
        "\n",
        "    plt.subplot(1, 2, 2)\n",
        "    sns.scatterplot(x=embedding[:, 0], y=embedding[:, 1], hue=labels, palette='coolwarm', s=10)\n",
        "    plt.title(f'UMAP projection of the shared feature space (colored by vital status) {title_suffix}')\n",
        "\n",
        "    plt.tight_layout()\n",
        "    torch.cuda.synchronize()\n",
        "    plt.show()\n",
        "\n",
        "# Función para visualizar UMAP de una capa específica en una época e iteración específica\n",
        "def visualize_layer_umap(epoch, iteration, module_name, layer_name, title_suffix=''):\n",
        "    features, labels, domains = load_layer_features_from_disk(epoch, iteration, module_name, layer_name)\n",
        "    if features is not None:\n",
        "        reducer = umap.UMAP(n_components=2)\n",
        "        embedding = reducer.fit_transform(features)\n",
        "        plot_umap(embedding, labels, domains, title_suffix)\n",
        "\n",
        "# Parámetros\n",
        "input_dim = X.shape[1]\n",
        "num_domains = len(label_encoder_panel.classes_)\n",
        "num_epochs = 10\n",
        "\n",
        "# Crear el modelo\n",
        "model = DANN(input_dim, num_domains).to(device)\n",
        "\n",
        "# Imprimir el resumen de la arquitectura del modelo usando torchsummary\n",
        "summary(model, input_size=(input_dim,))\n",
        "\n",
        "# Después de llamar a summary, puedes ajustar el lambda_ durante el entrenamiento\n",
        "lambda_ = 0.2 # Valor ajustado por el usuario\n",
        "\n",
        "# Definir el optimizador AdamW\n",
        "optimizer = optim.AdamW(\n",
        "    model.parameters(),\n",
        "    lr=0.01,  # Tasa de aprendizaje\n",
        "    betas=(0.9, 0.99),  # Coeficientes beta\n",
        "    eps=1e-8,  # Valor pequeño para estabilidad numérica\n",
        "    weight_decay=0.01,  # Regularización L2\n",
        "    amsgrad=False  # Si deseas usar la variante AMSGrad\n",
        ")\n",
        "\n",
        "# Definir las funciones de pérdida\n",
        "criterion_label = nn.BCELoss().to(device)\n",
        "criterion_domain = nn.CrossEntropyLoss().to(device)\n",
        "\n",
        "# Listas para guardar las precisiones\n",
        "label_accuracies = []\n",
        "domain_accuracies = []\n",
        "\n",
        "\n",
        "class ModelSaver:\n",
        "    def __init__(self):\n",
        "        self.models = []\n",
        "\n",
        "    def save_model(self, model, iteration):\n",
        "        # Hacer una copia del estado del modelo\n",
        "        model_copy = {k: v.clone() for k, v in model.state_dict().items()}\n",
        "        self.models.append((iteration, model_copy))\n",
        "\n",
        "    def get_models(self):\n",
        "        return self.models\n",
        "# Crear instancia de ModelSaver\n",
        "model_saver = ModelSaver()\n",
        "\n",
        "\n",
        "\n",
        "# Entrenamiento y guardado de características\n",
        "for epoch in range(num_epochs):\n",
        "    model.train()\n",
        "    correct_label = 0\n",
        "    correct_domain = 0\n",
        "    total = 0\n",
        "\n",
        "    for i, (inputs, labels, domains) in enumerate(loader):\n",
        "        inputs, labels, domains = inputs.to(device), labels.to(device), domains.to(device)\n",
        "\n",
        "        # Forward pass: obtiene las predicciones para etiquetas y dominios\n",
        "        label_preds, domain_preds, features = model(inputs, lambda_)  # El valor de lambda_ es proporcionado aquí\n",
        "\n",
        "        # Cálculo de pérdidas\n",
        "        loss_label = criterion_label(label_preds, labels)  # Pérdida de predicción de etiquetas (L_y)\n",
        "        loss_domain = criterion_domain(domain_preds, domains)  # Pérdida de predicción de dominios (L_d)\n",
        "\n",
        "        # Reseteo de gradientes acumulados\n",
        "        optimizer.zero_grad()\n",
        "\n",
        "        # Suma de las pérdidas de ambas tareas\n",
        "        loss = loss_label #+ loss_domain\n",
        "\n",
        "        # Retropropagación de gradientes (aquí es donde ocurren ambas retropropagaciones)\n",
        "        loss.backward()\n",
        "\n",
        "        # Clip gradients to prevent explosion\n",
        "        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)\n",
        "\n",
        "        # Actualización de los parámetros del modelo\n",
        "        optimizer.step()\n",
        "\n",
        "        predicted_label = (label_preds > 0.5).float()\n",
        "        correct_label += (predicted_label == labels).sum().item()\n",
        "\n",
        "        _, predicted_domain = torch.max(domain_preds, 1)\n",
        "        correct_domain += (predicted_domain == domains).sum().item()\n",
        "\n",
        "        total += labels.size(0)\n",
        "\n",
        "        # Guardar características solo para la última iteración dentro de cada época\n",
        "        if i == len(loader) - 1:  # Solo para la última iteración\n",
        "            save_layer_features_to_disk((epoch, i), model, loader, 'feature_extractor', 'fc1')\n",
        "            save_layer_features_to_disk((epoch, i), model, loader, 'domain_classifier', 'fc2', use_features=True)\n",
        "            save_layer_features_to_disk((epoch, i), model, loader, 'label_predictor', 'fc2', use_features=True)\n",
        "\n",
        "\n",
        "\n",
        "            # Guardar el modelo después de cada iteración (en este caso la ulitima iteración)\n",
        "            model_saver.save_model(model, (epoch, i))\n",
        "\n",
        "\n",
        "\n",
        "    label_accuracy = correct_label / total\n",
        "    domain_accuracy = correct_domain / total\n",
        "\n",
        "    label_accuracies.append(label_accuracy)\n",
        "    domain_accuracies.append(domain_accuracy)\n",
        "\n",
        "    print(f'Epoch [{epoch+1}/{num_epochs}], Label Accuracy: {label_accuracy:.4f}, Domain Accuracy: {domain_accuracy:.4f}')\n",
        "\n",
        "# Visualización de UMAP de la última iteración de la segunda época (indexada como epoch 1)\n",
        "visualize_layer_umap(epoch=1, iteration=len(loader) - 1, module_name='feature_extractor', layer_name='fc1', title_suffix='for layer fc1 in FeatureExtractor at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=1, iteration=len(loader) - 1, module_name='domain_classifier', layer_name='fc2', title_suffix='for layer fc2 in DomainClassifier at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=1, iteration=len(loader) - 1, module_name='label_predictor', layer_name='fc2', title_suffix='for layer fc2 in LabelPredictor at epoch 2, last iteration')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lhrVLsxk6u_v"
      },
      "outputs": [],
      "source": [
        "# Visualización de UMAP de la última iteración de la segunda época (indexada como epoch 1)\n",
        "visualize_layer_umap(epoch=4, iteration=len(loader) - 1, module_name='feature_extractor', layer_name='fc1', title_suffix='for layer fc2 in FeatureExtractor at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=4, iteration=len(loader) - 1, module_name='domain_classifier', layer_name='fc2', title_suffix='for layer fc2 in DomainClassifier at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=4, iteration=len(loader) - 1, module_name='label_predictor', layer_name='fc2', title_suffix='for layer fc2 in LabelPredictor at epoch 2, last iteration')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "toDFc0wn6xQG"
      },
      "outputs": [],
      "source": [
        "# Visualización de UMAP de la última iteración de la segunda época (indexada como epoch 1)\n",
        "visualize_layer_umap(epoch=9, iteration=len(loader) - 1, module_name='feature_extractor', layer_name='fc1', title_suffix='for layer fc2 in FeatureExtractor at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=9, iteration=len(loader) - 1, module_name='domain_classifier', layer_name='fc2', title_suffix='for layer fc2 in DomainClassifier at epoch 2, last iteration')\n",
        "visualize_layer_umap(epoch=9, iteration=len(loader) - 1, module_name='label_predictor', layer_name='fc2', title_suffix='for layer fc2 in LabelPredictor at epoch 2, last iteration')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "h_xkspDLzd3G"
      },
      "outputs": [],
      "source": [
        "# Supongamos que la época 5 es la que te interesa (indexada como 4 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 5 es epoch 4\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 5\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pX2Bkqzj4XH1"
      },
      "outputs": [],
      "source": [
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score, roc_curve\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Definir el número de pliegues\n",
        "n_splits = 10  # Número de pliegues para la validación cruzada\n",
        "skf = StratifiedKFold(n_splits=n_splits)\n",
        "\n",
        "# Listas para almacenar resultados de AUC\n",
        "auc_scores = []\n",
        "fpr_list = []\n",
        "tpr_list = []\n",
        "\n",
        "# Validación cruzada estratificada\n",
        "for fold, (train_index, test_index) in enumerate(skf.split(X, Y)):\n",
        "    print(f\"Fold {fold + 1}/{n_splits}\")\n",
        "\n",
        "    # Dividir los datos en entrenamiento y prueba\n",
        "    X_train, X_test = X[train_index], X[test_index]\n",
        "    Y_train, Y_test = Y[train_index], Y[test_index]\n",
        "\n",
        "    # Convertir a tensores\n",
        "    X_train_tensor = torch.tensor(X_train, dtype=torch.float32).to(device)\n",
        "    Y_train_tensor = torch.tensor(Y_train, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "    X_test_tensor = torch.tensor(X_test, dtype=torch.float32).to(device)\n",
        "    Y_test_tensor = torch.tensor(Y_test, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "\n",
        "    # Evaluar el modelo en el conjunto de prueba\n",
        "    model.eval()\n",
        "    with torch.no_grad():\n",
        "        label_preds, _, _ = model(X_test_tensor)\n",
        "        label_preds = label_preds.cpu().numpy()\n",
        "        Y_test = Y_test_tensor.cpu().numpy()\n",
        "\n",
        "    # Calcular el AUC\n",
        "    auc_score = roc_auc_score(Y_test, label_preds)\n",
        "    auc_scores.append(auc_score)\n",
        "\n",
        "    # Calcular la curva ROC\n",
        "    fpr, tpr, _ = roc_curve(Y_test, label_preds)\n",
        "    fpr_list.append(fpr)\n",
        "    tpr_list.append(tpr)\n",
        "\n",
        "    print(f\"AUC for fold {fold + 1}: {auc_score:.4f}\")\n",
        "\n",
        "# Promediar y graficar las curvas ROC\n",
        "plt.figure(figsize=(10, 8))\n",
        "for i in range(n_splits):\n",
        "    plt.plot(fpr_list[i], tpr_list[i], label=f'Fold {i+1} (AUC = {auc_scores[i]:.4f})')\n",
        "\n",
        "plt.plot([0, 1], [0, 1], 'k--', lw=2)\n",
        "plt.xlim([0.0, 1.0])\n",
        "plt.ylim([0.0, 1.0])\n",
        "plt.xlabel('False Positive Rate')\n",
        "plt.ylabel('True Positive Rate')\n",
        "plt.title('ROC Curves for Label Predictor across Folds')\n",
        "plt.legend(loc=\"lower right\")\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "X_tensor.shape"
      ],
      "metadata": {
        "id": "NwkAECYRkBKK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Analysis of the Model Outputs:**\n",
        "\n",
        "The result you obtained confirms that the wrapper (`ModelWrapper`) is functioning correctly to return only the first output of the model, which corresponds to the vital status prediction. Below is the analysis of the results:\n",
        "\n",
        "1. **Outputs of the model without the wrapper:**\n",
        "   - **First output:** `tensor([[4.6149e-02], ... , [6.9630e-02]], device='cuda:0')` with dimension `[10286, 1]` corresponds to the vital status prediction (`LabelPredictor`).\n",
        "   - **Second output:** `tensor([[-0.2760,  0.1515, ...]], device='cuda:0')` with dimension `[10286, 33]` corresponds to the domain prediction (`DomainClassifier`).\n",
        "   - **Third output:** `tensor([[ 2.3175e-01, ...]], device='cuda:0')` with dimension `[10286, 5000]` corresponds to some intermediate representation or features from the feature extractor (`FeatureExtractor`).\n",
        "\n",
        "2. **Output of the model with the wrapper:**\n",
        "   - **Output:** `tensor([[4.6149e-02], ... , [6.9630e-02]], device='cuda:0')` with dimension `[10286, 1]` corresponds to the first output of the model, which is the vital status prediction, as expected.\n",
        "\n",
        "**Conclusion:**\n",
        "\n",
        "The wrapper `ModelWrapper` is functioning as intended, returning only the first output of the full model, which is the vital status prediction. This is confirmed by comparing the dimensions:\n",
        "\n",
        "- **Dimensions of all outputs:**\n",
        "  - **First output:** `[10286, 1]`\n",
        "  - **Second output:** `[10286, 33]`\n",
        "  - **Third output:** `[10286, 5000]`\n",
        "\n",
        "- **Dimension of the wrapped output:**\n",
        "  - `[10286, 1]`\n",
        "\n",
        "The dimension `[10286, 1]` of the wrapped output matches the dimension of the vital status prediction, confirming that the wrapper is correctly returning the first output.\n",
        "\n",
        "This behavior ensures that when you calculate SHAP values using the wrapper, you are calculating SHAP values with respect to the vital status prediction (and not the other outputs of the model). This is exactly what you need to analyze how the input features influence the vital status prediction in your DANN model.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "P9f8Jf5GopPA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "asdasdasdasdasdasd"
      ],
      "metadata": {
        "id": "KVrWN6MitUQo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import shap\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score, roc_curve\n",
        "import matplotlib.pyplot as plt\n",
        "import torch\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo (CPU o GPU)\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "print(X_tensor.shape)\n",
        "\n",
        "# Supongamos que la época 10 es la que te interesa (indexada como 9 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 10 es epoch 9\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 9\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegúrate de que el modelo completo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Obtener la salida del modelo\n",
        "output = model(X_tensor)\n",
        "\n",
        "# Si el modelo devuelve una tupla, selecciona la primera salida\n",
        "if isinstance(output, tuple):\n",
        "    output = output[0]\n",
        "\n",
        "# Paso 1: Usar GradientExplainer para calcular los valores SHAP basados en las características originales\n",
        "explainer = shap.GradientExplainer(model, X_tensor)\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "print(shap_values.shape)\n",
        "\n",
        "# Paso 2: Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Paso 3: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "X1YyYkYQS5ZK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import shap\n",
        "import torch\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "import numpy as np\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo (CPU o GPU)\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "\n",
        "# Supongamos que la época 5 es la que te interesa (indexada como 4 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 10 es epoch 9\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 9\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegúrate de que el modelo completo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Paso 1: Crear un objeto DeepExplainer para el modelo\n",
        "explainer = shap.DeepExplainer(model.label_predictor, X_tensor)\n",
        "\n",
        "# Paso 2: Calcular los valores SHAP utilizando el tensor de entrada X_tensor\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "\n",
        "# Paso 3: Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Verifica las dimensiones de shap_values para asegurarte de que coincide con lo esperado\n",
        "print(\"Dimensión de shap_values:\", shap_values_np.shape)\n",
        "\n",
        "# Paso 4: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "Cy_w0XwU7aZ2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import shap\n",
        "import torch\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "import numpy as np\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo (CPU o GPU)\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "\n",
        "# Supongamos que la época 5 es la que te interesa (indexada como 4 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 10 es epoch 9\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 9\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegúrate de que el modelo completo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Verificar las salidas del modelo sin el wrapper\n",
        "with torch.no_grad():\n",
        "    all_outputs = model(X_tensor)\n",
        "    print(\"Salidas del modelo (sin wrapper):\", all_outputs)\n",
        "    print(\"Dimensiones de todas las salidas:\", [output.shape for output in all_outputs])\n",
        "\n",
        "# Crear una clase wrapper para el modelo\n",
        "# Esta clase `ModelWrapper` envuelve el modelo completo, pero en la función `forward`\n",
        "# devuelve solo la primera salida del modelo (que corresponde a la predicción del estado vital).\n",
        "class ModelWrapper(torch.nn.Module):\n",
        "    def __init__(self, model):\n",
        "        super(ModelWrapper, self).__init__()\n",
        "        self.model = model\n",
        "\n",
        "    def forward(self, x):\n",
        "        return self.model(x)[0]  # Retornar solo la primera salida (la predicción de la etiqueta de vitalidad)\n",
        "\n",
        "# Instanciar la clase wrapper\n",
        "wrapped_model = ModelWrapper(model)\n",
        "\n",
        "# Verificar la salida del modelo con el wrapper\n",
        "with torch.no_grad():\n",
        "    wrapped_output = wrapped_model(X_tensor)\n",
        "    print(\"Salida del modelo (con wrapper):\", wrapped_output)\n",
        "    print(\"Dimensión de la salida envuelta:\", wrapped_output.shape)\n",
        "\n",
        "# Paso 1: Usar GradientExplainer para calcular los valores SHAP basados en las características originales\n",
        "explainer = shap.DeepExplainer(wrapped_model, X_tensor)\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "\n",
        "# Paso 2: Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Verifica las dimensiones de shap_values para asegurarte de que coincide con lo esperado\n",
        "print(\"Dimensión de shap_values:\", shap_values_np.shape)\n",
        "\n",
        "# Paso 3: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "GoPFZx9FoKqv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n",
        "\n"
      ],
      "metadata": {
        "id": "b6Vfeqsp2hhN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)"
      ],
      "metadata": {
        "id": "eCnTIiCJ-3b9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "jGeUiZFR-3z4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import shap\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score, roc_curve\n",
        "import matplotlib.pyplot as plt\n",
        "import torch\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo (CPU o GPU)\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "\n",
        "# Supongamos que la época 5 es la que te interesa (indexada como 4 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 10 es epoch 9\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 9\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegúrate de que el modelo completo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Crear una clase wrapper para el modelo\n",
        "class ModelWrapper(torch.nn.Module):\n",
        "    def __init__(self, model):\n",
        "        super(ModelWrapper, self).__init__()\n",
        "        self.model = model\n",
        "\n",
        "    def forward(self, x):\n",
        "        return self.model(x)[0]  # Retornar solo la primera salida\n",
        "\n",
        "# Instanciar la clase wrapper\n",
        "wrapped_model = ModelWrapper(model)\n",
        "\n",
        "# Paso 1: Usar GradientExplainer para calcular los valores SHAP basados en las características originales\n",
        "explainer = shap.GradientExplainer(wrapped_model, X_tensor,batch_size = 150)\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "\n",
        "# Paso 2: Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Paso 3: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "ujUp07ZvqmgC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sdasdasdasdasdasdasdasd"
      ],
      "metadata": {
        "id": "DomOdD_IymHk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "onV5GlO2Br6b"
      },
      "outputs": [],
      "source": [
        "import shap\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score, roc_curve\n",
        "import matplotlib.pyplot as plt\n",
        "import torch\n",
        "\n",
        "# Preparar los datos\n",
        "X = df.iloc[:, :-2].values  # Datos de miRNAs\n",
        "Y = df['vital'].values  # Estado vital\n",
        "panels = df['panel'].values  # Etiquetas del dominio (paneles)\n",
        "\n",
        "# Convertir el estado vital a valores numéricos\n",
        "label_encoder_vital = LabelEncoder()\n",
        "Y = label_encoder_vital.fit_transform(Y)\n",
        "\n",
        "# Convertir los datos a tensores y moverlos al dispositivo (CPU o GPU)\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "X_tensor = torch.tensor(X, dtype=torch.float32).to(device)\n",
        "Y_tensor = torch.tensor(Y, dtype=torch.float32).unsqueeze(1).to(device)\n",
        "\n",
        "# Supongamos que la época 5 es la que te interesa (indexada como 4 en Python)\n",
        "epoch_of_interest = 9  # Indexada desde 0, por lo que la época 10 es epoch 9\n",
        "\n",
        "# Obtener el modelo guardado al final de la época 9\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "epoch, iteration = iteration_info\n",
        "\n",
        "print(f\"Modelo guardado al final de la época {epoch + 1}, iteración {iteration + 1}\")\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegúrate de que el modelo completo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Crear una clase wrapper para el modelo\n",
        "class ModelWrapper(torch.nn.Module):\n",
        "    def __init__(self, model):\n",
        "        super(ModelWrapper, self).__init__()\n",
        "        self.model = model\n",
        "\n",
        "    def forward(self, x):\n",
        "        return self.model(x)[0]  # Retornar solo la primera salida\n",
        "\n",
        "# Instanciar la clase wrapper\n",
        "wrapped_model = ModelWrapper(model)\n",
        "\n",
        "# Paso 1: Usar GradientExplainer para calcular los valores SHAP basados en las características originales\n",
        "explainer = shap.GradientExplainer(wrapped_model, X_tensor)\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "\n",
        "# Paso 2: Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Paso 3: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import shap\n",
        "import torch\n",
        "\n",
        "# Asumiendo que el modelo DANN ya ha sido definido y entrenado\n",
        "\n",
        "# Instanciar y cargar el modelo como se hace en el código original\n",
        "input_dim = X.shape[1]\n",
        "num_domains = len(label_encoder_panel.classes_)\n",
        "model = DANN(input_dim, num_domains).to(device)\n",
        "\n",
        "# Cargar el modelo guardado al final de una época en particular\n",
        "epoch_of_interest = 9  # Ejemplo, puedes ajustar la época de interés\n",
        "iteration_info, model_state_dict = model_saver.get_models()[epoch_of_interest]\n",
        "model.load_state_dict(model_state_dict)\n",
        "\n",
        "# Asegurarse de que el modelo esté en modo de evaluación\n",
        "model.eval()\n",
        "\n",
        "# Crear una clase wrapper para obtener solo la salida del LabelPredictor\n",
        "class LabelPredictorWrapper(torch.nn.Module):\n",
        "    def __init__(self, model):\n",
        "        super(LabelPredictorWrapper, self).__init__()\n",
        "        self.label_predictor = model.label_predictor\n",
        "\n",
        "    def forward(self, x):\n",
        "        return self.label_predictor(x)\n",
        "\n",
        "# Instanciar el wrapper\n",
        "label_predictor_model = LabelPredictorWrapper(model)\n",
        "\n",
        "# Usar GradientExplainer para calcular los valores SHAP basados en las características originales\n",
        "explainer = shap.GradientExplainer(label_predictor_model, X_tensor)\n",
        "shap_values = explainer.shap_values(X_tensor)\n",
        "\n",
        "# Convertir los valores SHAP a numpy para visualización\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Visualización de SHAP\n",
        "shap.summary_plot(shap_values_np, X, feature_names=df.iloc[:, :-2].columns)\n"
      ],
      "metadata": {
        "id": "ghUSogvJnBZg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dec4KHkDP1TB"
      },
      "outputs": [],
      "source": [
        "\n",
        "# Paso 2: Usar GradientExplainer para calcular los valores SHAP\n",
        "explainer = shap.GradientExplainer(label_predictor_model,X)\n",
        "shap_values = explainer.shap_values(X)\n",
        "\n",
        "# Paso 3: Convertir los tensores de entrada y los valores SHAP a numpy para visualización\n",
        "features_np = X.cpu().numpy()\n",
        "shap_values_np = np.array(shap_values).squeeze()\n",
        "\n",
        "# Paso 4: Visualización de SHAP con feature_names tomadas del DataFrame original\n",
        "shap.summary_plot(shap_values_np, features_np, feature_names=df.iloc[:, :-2].columns)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RrMs6EmsOntT"
      },
      "outputs": [],
      "source": [
        "features_np.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cF4NrvEHOdTY"
      },
      "outputs": [],
      "source": [
        "len(df.iloc[:, :-2].columns)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dNkEpQpHIwIg"
      },
      "outputs": [],
      "source": [
        "shap.summary_plot(shap_values_np, features_np, feature_names=df.iloc[:, :-2].columns,max_display=50 )\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WxyfCYZ3OXnh"
      },
      "outputs": [],
      "source": [
        "shap_values_np.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PEbS7RgTIUQo"
      },
      "outputs": [],
      "source": [
        "\n",
        "pd.DataFrame(shap_values_np, columns=df.iloc[:, :-2].columns,index=df.index)"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "L4",
      "machine_shape": "hm",
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}